### PR TITLE
ENH: New type called `AMRFinderPlusDatabase`

### PR DESCRIPTION
**Pull request by VinzentRisch**
solves #80 

This new type is used for the storage of the database for the amrfinderplus tool.

### Set up an environment 
https://github.com/bokulich-lab/q2-amr?tab=readme-ov-file#installation

### Run it locally 
1. First, clone the repo and checkout the PR branch:
```bash
git clone https://github.com/bokulich-lab/q2-amr.git
cd q2-amr
git fetch origin pull/81/head:pr-81
git checkout pr-81
pip install -e .
```
2. Download test files 
amrfinderplus_database.zip: https://polybox.ethz.ch/index.php/s/nJmbGpkM1ywS9VW

3. Test it out!
```bash
qiime tools import --input-path database --output-path database.qza --type AMRFinderPlusDatabase 
```

### CODE REVIEW

**Comment by VinzentRisch** [comment]
Hey @misialq 
Could you have a look at this? I now created two formats that just validate any binary or text file. Should I add formats for the tab separated files? 
I could add new formats for these files that just validates the header:
AMR_DNA-Streptococcus_pneumoniae.tab
changes.txt
database_format_version.txt
fam.tab
taxgroup.tab
version.txt
AMRProt-mutation.tab
AMRProt-susceptible.tab

Those would be 8 new formats so Im not so sure if that even makes sense.

File: q2_amr/amrfinderplus/types/_format.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    def _validate_(self, level):
+        pass
+
+
+class BinaryFormat(model.BinaryFileFormat):
+    def _validate_(self, level):
+        pass
+
+
+class AMRFinderPlusDatabaseDirectoryFormat(model.DirectoryFormat):
```
Comment (misialq) [inline q2_amr/amrfinderplus/types/_format.py:22]: OK, so this is a little bit too wild 😅 What I propose here is:
1. rename all the variable names to by all lowercase - there is really no reason for them to reflect the file names exactly 1-to-1 and, by convention, we use all-uppercase to represent constants
2. you should group files with the same extension into `FileCollection`s - just adjust the regex for the file name so that it can recognize files for all the different organisms
3. I'm not sure whether it makes sense to have separate validation for all the tab-separated files? Do you find any of those particularly important? 😅

File: q2_amr/amrfinderplus/types/_transformer.py
Comment (misialq) [inline q2_amr/amrfinderplus/types/_transformer.py:1]: Why is this file here? If you don't intend to add any transformers you can remove it.

File: q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii
Comment (misialq) [inline q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii:1]: Once you change the directory format as I pointed out above, you can remove most of those test files - they are just polluting the repo.

**Review by misialq — CHANGES_REQUESTED** [review]

File: q2_amr/amrfinderplus/types/_format.py (hunk 1 above)
Comment (VinzentRisch) [reply q2_amr/amrfinderplus/types/_format.py:22]: First I also had the `FileCollection` but then there is no way to verify if all needed files are present.
I need 11 files with the extension nto for instance and with `FileCollection` it would be possible to import a database that has any number of files with the extension nto. Or is there a way to specify how many files there should be in one `FileCollection`?

And not really, they are all important.

File: q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii
Comment (VinzentRisch) [reply q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii:1]: I left them there, because if I remove most files, the test would become meaningless. 
Only like this it is actually tested if all regex expressions work with the database files.

**Comment by VinzentRisch** [comment]
Hi @misialq could you have another look at this?
I'm not so sure about the path makers. Do they work like this with a regex?

File: q2_amr/amrfinderplus/types/_format.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    )
+    amr_dna_comp = model.FileCollection(
+        r"^AMR_DNA-[a-zA-Z_]+\.n..$", format=BinaryFormat
+    )
+    amr_dna_tab = model.FileCollection(r"^AMR_DNA-[a-zA-Z_]+\.tab$", format=TextFormat)
+    amr_cds_comp = model.FileCollection(r"^AMR_CDS\.n..$", format=BinaryFormat)
+    amr_cds = model.File("AMR_CDS", format=MixedCaseDNAFASTAFormat)
+
+    @amr_lib_comp.set_path_maker
+    def amr_lib_comp_path_maker(self):
```
Comment (misialq) [inline q2_amr/amrfinderplus/types/_format.py:46]: Hey @Vinzent, I don't think that's how they work - they don't need to be regular expressions. The should be expressions that will pretty much just construct a file name for the given file - you can find some more illustrative examples in `q2_types/per_sample_sequences/_format.py`.

Comment (vinzent) [reply q2_amr/amrfinderplus/types/_format.py:46]: you marked the wrong vinzent. ;-)

Comment (VinzentRisch) [reply q2_amr/amrfinderplus/types/_format.py]: @misialq Okay I changed it.

File: q2_amr/amrfinderplus/types/_format.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    amrprot = model.File("AMRProt", format=ProteinFASTAFormat)
+    amrprot_blast = model.FileCollection(r"^AMRProt\.p..$", format=BinaryFormat)
+    amrprot_mutation = model.File("AMRProt-mutation.tab", format=TextFormat)
+    amrprot_suppress = model.File("AMRProt-suppress", format=TextFormat)
+    amrprot_susceptible = model.File("AMRProt-susceptible.tab", format=TextFormat)
+    changes = model.File("changes.txt", format=TextFormat)
+    db_version = model.File("database_format_version.txt", format=TextFormat)
+    fam = model.File("fam.tab", format=TextFormat)
+    taxgroup = model.File("taxgroup.tab", format=TextFormat)
+    version = model.File("version.txt", format=TextFormat)
```
Comment (misialq) [inline q2_amr/amrfinderplus/types/_format.py:34]: I'm wondering here: what's the difference between the contents of `db_version` and `version`? Also is there anything interesting in the `changes`? Technically, we would not need to maintain all of these files. Maybe just a single file with version would be enough?

File: q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii
Comment (misialq) [reply q2_amr/amrfinderplus/types/tests/data/database/AMR_DNA-Acinetobacter_baumannii:1]: Why do think it will become meaningless? Your test should just make sure that the regex is working - for this you don't need all the files for all the species - you could just keep a set of those for, say, two taxa and remove everything else (now that you have the FileCollection in place).

File: q2_amr/amrfinderplus/types/_format.py (hunk 3 above)
Comment (VinzentRisch) [reply q2_amr/amrfinderplus/types/_format.py:34]: i just thought it would be easier to leave everything in because the files are very small anyway. But I can also take them out.
actually the AMR_CDS files are also not needed, so I'll remove them too.

Comment (VinzentRisch) [reply q2_amr/amrfinderplus/types/_format.py]: Ah yes i forgot to change it in the PR test files but yes that is how it should be. :) 
When you approve, can you also merge for me. Because those required checks are still not gone.

**Review by misialq — APPROVED** [review]
Looks good! 🚀
